astropy/io/fits/diff.py

```diff
@@ -317,11 +331,20 @@ def __init__(self, a, b, ignore_keywords=[], ignore_comments=[],
                 b.close()
 
     def _diff(self):
+        if self.ignore_hdus:
+            self.a = [h for h in self.a if h.name not in self.ignore_hdus]
+            self.b = [h for h in self.b if h.name not in self.ignore_hdus]
+        if self.ignore_hdu_patterns:
+            for pattern in self.ignore_hdu_patterns:
+                self.a = [h for h in self.a if not fnmatch.fnmatch(h.name, pattern)]
+                self.b = [h for h in self.b if not fnmatch.fnmatch(h.name, pattern)]
```

Comment:
You could maybe build a list of HDU names to avoid looping several times on the HDUList (this would also also to use `fnmatch.filter`, as for keywords).
Also, for consistency with keywords, `diff_hdu_count` should probably be computed before filtering HDUs ? I think it makes sense to always report that the number of HDUs is different.

Reply:
Thanks for the suggestion to use `fnmatch.filter`.  I learned something new today.  I have pushed a new commit using this, and also correcting an oversight in my list comps not actually being `HDUList` instances in the assignment to `self.a` and `self.b`.  Ack!

As for `diff_hdu_count`, if I compute it before the filtering, then it will show up as a diff if there are multiple ignored extensions with NAME in one vs the other, and the final assert in my test will fail.  I.e. it won't allow the ignoring of EXTVER.  Is this the desired functionality?

Reply:
I think so (for `diff_hdu_count`), not completely sure though, but there is the same behavior for `ignore_keywords`:
```
In [12]: h = fits.Header(dict(a=1,b=2.3,c='FOO'))

In [13]: h2 = fits.Header(dict(a=1,b=2.3,c='FOO',d='BAZ'))

In [14]: print(fits.HeaderDiff(h,h2).report())
 Headers have different number of cards:
  a: 3
  b: 4
 Extra keyword 'D'      in b: 'BAZ'

In [15]: print(fits.HeaderDiff(h,h2,ignore_keywords=['D']).report())
 Headers have different number of cards:
  a: 3
  b: 4

In [16]: print(fits.HeaderDiff(h,h2,ignore_keywords=['D']).identical)
False
```

Reply:
I've fixed the behavior so that it will now still report a `diff_hdu_count` even if one has passed `ignore_hdus=True`.